### PR TITLE
Movies repository

### PR DESCRIPTION
**Pull request by opsat7373**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -39,6 +43,10 @@ dependencies {
     implementation 'androidx.appcompat:appcompat:1.2.0'
     implementation 'com.google.android.material:material:1.2.1'
     implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
```
Comment (ArtemKleinschmidt) [inline app/build.gradle]: What the purpose of this dependency?

File: app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import com.opsat7373.moviestars.R
+import com.opsat7373.moviestars.databinding.MovieItemBinding
+import com.opsat7373.moviestars.data.model.Movie
+import java.util.*
+
+class MovieItemAdapter(moviesList : List<Movie?>?) : RecyclerView.Adapter<MovieItemAdapter.MovieItemViewHolder>() {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt]: looks like passing data to constructor is useless, as you update it with setList() method

File: app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    var moviesList : List<Movie?>
+
+    fun setList(moviesList : List<Movie?>?) {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt]: The list and the parameter should not be nullable

File: app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        return MovieItemViewHolder(itemView)
+    }
+
+    override fun onBindViewHolder(holder: MovieItemViewHolder, position: Int) {
+        holder.binding.movieName.text = moviesList[position]?.original_title.toString()
+        println("position : ${position.toString()}")
+    }
+
+    override fun getItemCount(): Int {
+        println(moviesList.size)
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt]: use logcat for logs

Comment (ArtemKleinschmidt) [reply app/src/main/java/com/opsat7373/moviestars/adapters/MovieItemAdapter.kt]: Log.d(TAG, "logmessage")

File: app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.*
+
+
+class WebDataSource : MovieDataSourceInterface {
+    override fun getAll(): MutableLiveData<List<Movie?>> {
+        val retrofit = Retrofit.Builder()
+            .baseUrl("https://api.themoviedb.org/")
+            .addConverterFactory(GsonConverterFactory.create())
+            .build()
+        val service = retrofit.create(TMDBService::class.java)
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt]: Pass the service as an external dependency using a constructor.
No need to build a retrofit client for each call.

File: app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import java.util.*
+
+
+class WebDataSource : MovieDataSourceInterface {
+    override fun getAll(): MutableLiveData<List<Movie?>> {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt]: If you do not modify the list, it's better to expose just LiveData, not Mutable

File: app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        val retrofit = Retrofit.Builder()
+            .baseUrl("https://api.themoviedb.org/")
+            .addConverterFactory(GsonConverterFactory.create())
+            .build()
+        val service = retrofit.create(TMDBService::class.java)
+        val request: Call<PopularMoviesList> = service.getPopularList()
+
+
+        var data :  MutableLiveData<List<Movie?>> =  MutableLiveData<List<Movie?>>()
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt]: check if you can return non-null values

File: app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        request.enqueue(object : Callback<PopularMoviesList> {
+            override fun onResponse(call: Call<PopularMoviesList>?,
+                           response: Response<PopularMoviesList>) {
+                if (response.isSuccessful()) {
+                    data.value = response.body()?.results ?: LinkedList<Movie>()
+                }
+            }
+
+            override fun onFailure(call: Call<PopularMoviesList>?, t: Throwable?) {
+                data.value = LinkedList<Movie>()
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt]: add error log, using logcat

File: app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package com.opsat7373.moviestars.data.model
+private val moviesList : Array<String> =  arrayOf("Die hard 1" , "2012", "Star wars")
+private val moviesList2 : Array<String> =  arrayOf("Die hard 2" , "2012 2", "Star wars 2")
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt]: use ArrayList instead

File: app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+data class Movie(val original_title : String) {
+    companion object {
+        fun getMoviesList() : Array<Movie?> {
+            val moviesArray : Array<Movie?> = arrayOfNulls(moviesList.size)
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt]: use non-nullable values when it's possible.

`val moviesArray = ArrayList<Movie>()`

Comment (ArtemKleinschmidt) [reply app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt]: the method should return abstract class 
`List<Movie>`

File: app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package com.opsat7373.moviestars.data.model
+private val moviesList : Array<String> =  arrayOf("Die hard 1" , "2012", "Star wars")
+private val moviesList2 : Array<String> =  arrayOf("Die hard 2" , "2012 2", "Star wars 2")
+
+data class Movie(val original_title : String) {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt:3]: keep you data structures and data sources separately.
move data class to a separate file.

File: app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    companion object {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/model/Movie.kt]: no need for companion object here

File: app/src/main/java/com/opsat7373/moviestars/presentation/fragment/MoviesListFragment.kt [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        viewBinding = FragmentMoviesListFramgentBinding.inflate(layoutInflater)
+
+        val linearLayoutManager = LinearLayoutManager(activity)
+        moviesListViewModel.loadMoviesList()
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/presentation/fragment/MoviesListFragment.kt]: use generic name, like `init()`, as a ViewModel might do a lot of things during initialization, not just data loading

File: app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package com.opsat7373.moviestars.presentation.viewmodel
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.opsat7373.moviestars.data.model.Movie
+import com.opsat7373.moviestars.data.repository.MovieRepository
+
+class MoviesListViewModel : ViewModel() {
+    var moviesList = MutableLiveData<List<Movie?>>()
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt]: not-mutable live data is preferred.
if you need Mutable, you can keep private field as mutable, but expose just LiveData via public field.
`
private val _moviesList = MutableLiveData<List<Movie>>()
`
`
val movieList: LiveData<List<Movie>> = _movieList
`

File: app/src/main/res/layout/fragment_movies_list_framgent.xml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+        android:layout_marginTop="4dp"
+        android:text="@string/refresh_list_button"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/movies_list_recycler_view"
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_marginStart="12dp"
```
Comment (ArtemKleinschmidt) [inline app/src/main/res/layout/fragment_movies_list_framgent.xml]: keep your values in dimens file

File: app/src/main/res/layout/movie_item.xml [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content">
+
+    <TextView
+        android:id="@+id/movie_name"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:textSize="24sp"
```
Comment (ArtemKleinschmidt) [inline app/src/main/res/layout/movie_item.xml]: move to dimens file

File: app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt (hunk 6 above)
Comment (opsat7373) [reply app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt:15]: I am going modify this list in another methods. Also live data it's abstract class

File: app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt [hunk 17]
```diff
@@ -1,18 +1,25 @@
...
+        return this.getList()
     }
 
-    fun refreshList() {
-        moviesList  = MovieRepository().getMoviesList()
+    fun getList() : List<Movie> {
+        return moviesList.value ?: LinkedList()
+    }
+
+    fun getLiveData() : LiveData<List<Movie>> {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt]: It's better to use Kotlin backing field approach. 
So you have two fields, one is private, and the second is public, which returns a value from the private field

File: app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt [hunk 18]
```diff
@@ -1,18 +1,25 @@
...
 
-    fun loadMoviesList() {
+    fun initList() : List<Movie> {
         moviesList = MovieRepository().getMoviesList()
+        return this.getList()
     }
 
-    fun refreshList() {
-        moviesList  = MovieRepository().getMoviesList()
+    fun getList() : List<Movie> {
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt]: why do you need this method? 🤔

File: app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt [hunk 19]
```diff
@@ -10,28 +12,33 @@ import retrofit2.Retrofit
...
         val retrofit = Retrofit.Builder()
-            .baseUrl("https://api.themoviedb.org/")
-            .addConverterFactory(GsonConverterFactory.create())
-            .build()
-        val service = retrofit.create(TMDBService::class.java)
-        val request: Call<PopularMoviesList> = service.getPopularList()
+                .baseUrl("https://api.themoviedb.org/")
+                .addConverterFactory(GsonConverterFactory.create())
+                .build()
+        service = retrofit.create(TMDBService::class.java)
```
Comment (ArtemKleinschmidt) [inline app/src/main/java/com/opsat7373/moviestars/data/datasource/WebDataSource.kt]: extract retrofit client creation to a separate class

File: app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt (hunk 18 above)
Comment (opsat7373) [reply app/src/main/java/com/opsat7373/moviestars/presentation/viewmodel/MoviesListViewModel.kt:18]: i use this method for creation recycleView Adapter

**Review by ArtemKleinschmidt — APPROVED** [review]

**Pull request merged**